From e3e6d490ab90d28503debdc6b9d5084710f84dad Mon Sep 17 00:00:00 2001
From: Nicolas Cavallari <nicolas.cavallari@green-communications.fr>
Date: Fri, 27 Sep 2024 10:53:17 +0200
Subject: [PATCH] wifi: mt76: mt7915: Fix mesh scan on MT7916 DBDC

commit c4f075582304 ("wifi: mt76: mt7915: fix command timeout in AP stop
period") changes the behavior of mt7915_bss_info_changed() in mesh mode
when enable_beacon becomes false: it calls mt7915_mcu_add_bss_info(...,
false) and mt7915_mcu_add_sta(..., false) while the previous code
didn't.  These sends mcu commands that apparently confuse the firmware.

This breaks scanning while in mesh mode on AsiaRF MT7916 DBDC-based cards:
scanning works but no mesh frames get sent afterwards and the firmware
seems to be hosed.  It breaks on MT7916 DBDC but not on MT7915 DBDC.

Fixes: c4f075582304 ("wifi: mt76: mt7915: fix command timeout in AP stop period")
Signed-off-by: Nicolas Cavallari <nicolas.cavallari@green-communications.fr>
Link: https://patch.msgid.link/20240927085350.4594-1-nicolas.cavallari@green-communications.fr
Signed-off-by: Felix Fietkau <nbd@nbd.name>

[Neanderfunk: Backport fuer OpenWrt 23.05 / Gluon v2023.2.x, mt76
 2024-04-03 1e336a8. Upstream in mt76 seit 27.09.2024, in OpenWrt 24.10 /
 Gluon 2025.1 enthalten. Befund 2026-09-29: MERCUSYS MR90X (MT7986) unter
 Gluon 2023.2 - nach jedem WLAN-Scan ist Unicast ueber das Mesh des
 gescannten Radios tot, Broadcast und batman-OGMs laufen weiter, bis zum
 naechsten WLAN-Neustart (20 von 20); unter 2025.1.3 0 von 20. Ursache
 vermutlich wie unten: beim Scan schaltet mac80211 das Beaconing ab, der
 Treiber loescht dabei BSS- und Stations-Eintrag des Mesh-Interfaces in der
 Firmware. Am Geraet NICHT verifiziert - zum Test angeboten.]
---
 mt7915/main.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mt7915/main.c b/mt7915/main.c
index b16a6336..eaa47a9d 100644
--- a/mt7915/main.c
+++ b/mt7915/main.c
@@ -627,8 +627,9 @@ static void mt7915_bss_info_changed(struct ieee80211_hw *hw,
 	if (changed & BSS_CHANGED_ASSOC)
 		set_bss_info = vif->cfg.assoc;
 	if (changed & BSS_CHANGED_BEACON_ENABLED &&
+	    info->enable_beacon &&
 	    vif->type != NL80211_IFTYPE_AP)
-		set_bss_info = set_sta = info->enable_beacon;
+		set_bss_info = set_sta = 1;
 
 	if (set_bss_info == 1)
 		mt7915_mcu_add_bss_info(phy, vif, true);
